HULKs/hulk
# Twix Connection Status

## Introduced Changes

Fixes inconsistent connection state in twix.
The issue was that connection state changes were pushed into a bounded mpsc channel which in some cases would overflow.
Additionally, due to incorrect error handling, a full channel was handled the same way as a closed channel and no further updates would be sent.

This PR changes the way connection state and update notifications are propagated from the communication client through the `Nao` object to twix.
Instead of `mpsc` channels, `broadcast` and `watch` channels are used instead.
Broadcast channels abstract away our previous manual subscription concept.
Watch channels have the advantage of allowing repeated sync reads and remove the need to manually read from a channel and store the latest value somewhere.

Fixes #

## ToDo / Known Issues


## Ideas for Next Iterations (Not This PR)

See if watch & broadcast channels could be used in other places too.

## How to Test

Connect to nao, restart hulk service on nao.
Previously, this would result in an eternal yellow "connecting" status even after a connection had been reestablished.
With these changes, it should return back to a green "connected" after the service restarts and twix connects again.
